device/iot-device-samples/device-connection-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/DeviceClientManager.java

```diff
@@ -0,0 +1,415 @@
+            {
+                System.out.println("Failed to open the device client due to service throttling: " + e.getMessage());
+            }
+            catch (IotHubClientInternalServerErrorException e)
+            {
+                System.out.println("Failed to open the device client due to service encountering a transient error: " + e.getMessage());
+            }
+            catch (IotHubClientException e)
+            {
+                if (e.isRetryable())
+                {
+                    System.out.println("Failed to open the device client due to a retryable exception" + e.getMessage());
+                }
+                else
+                {
+                    System.out.println("Failed to open the device client due to a non-retryable exception" + e.getMessage());
+                    return false;
+                }
+            }
+            catch (TimeoutException e)
+            {
+                System.out.println("Timed out waiting for methods/twins subscription to be acknowledged by service, trying again.");
+            }
+
+            System.out.println("Sleeping a bit before retrying to open device client");
+            Thread.sleep(1000);
+        }
+    }
+
+    /**
+     * Sends telemetry and updates reported properties periodically. Returns if the client's connection was lost during this.
+     *
+     * This method simulates typical useage of a client, but can be modified to fit your use case without losing the
+     * automatic reconnection and retry that this sample has.
+     *
+     * @throws IllegalStateException If the client's connection was lost as this method attempted to send telemetry or
+     * update reported properties.
+     */
+    private void doWork() throws IllegalStateException
+    {
+        while (!reconnectNeeded)
+        {
+            sendTelemetryAsync();
+            updateReportedPropertiesAsync();
+
+            try
+            {
+                // not recommended for actual applications, but this makes the sample's logs print at a readable pace
+                Thread.sleep(10000);
+            }
+            catch (InterruptedException e)
+            {
+                System.out.println("Worker thread interrupted, likely because the sample is being stopped.");
+                return;
+            }
+        }
+    }
+
+    // region telemetry
+    private void sendTelemetryAsync()
+    {
+        Message messageToSend;
+        if (this.telemetryToResend.isEmpty())
+        {
+            // if no previous messages failed to send, send a new message
+            messageToSend = new Message("hello world");
+        }
+        else
+        {
+            // if any previous message failed to send, retry sending it before moving on to new messages
+            messageToSend = this.telemetryToResend.remove();
+        }
+
+        try
+        {
+            this.deviceClient.sendEventAsync(messageToSend, this, messageToSend);
```

Comment:
Why is `messageToSend` included twice? I'm guessing for the callback to be added to `telemetryToSend` on failure? Is that worth a code comment about this sample's strategy?